### ISSUE TITLE
[Stochastic RSI策略] | Stochastic RSI Strategy

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4486]
**标签**: quantitative-strategy
### 摘要
结合RSI与随机指标的优点，在超买超卖区域捕捉趋势反转点，实现高胜率的均值回归交易策略。

### 算法描述
**核心原理**

传统RSI在横盘市场表现良好，但在趋势行情中容易出现钝化；Stochastic Oscillator在趋势市场反应灵敏，但假信号较多。Stochastic RSI通过计算RSI的随机指标，将两者优势结合，产生更敏感的买卖信号。

**关键数学公式**

1. **RSI计算**：
$$RSI_t = 100 - \frac{100}{1 + RS_t}$$
$$RS_t = \frac{\sum_{i=1}^{n} \text{gain}_i}{\sum_{i=1}^{n} |\text{loss}_i|}$$

2. **Stochastic RSI（STOCHRSI）**：
$$STOCHRSI_t = \frac{RSI_t - \min(RSI_{t-n:t})}{\max(RSI_{t-n:t}) - \min(RSI_{t-n:t})}$$

3. **D通道（平滑K线）**：
$$K_t = SMA(STOCHRSI_t, n) = \frac{1}{n}\sum_{i=0}^{n-1} STOCHRSI_{t-i}$$

4. **D通道与K交叉**：
$$Signal = \begin{cases} \text{BUY} & \text{if } K_{t-1} < D_{t-1} \text{ and } K_t > D_t \text{ and } K_t < 0.2 \\ \text{SELL} & \text{if } K_{t-1} > D_{t-1} \text{ and } K_t < D_t \text{ and } K_t > 0.8 \\ \text{HOLD} & \text{otherwise} \end{cases}$$

5. **多周期确认**：
$$\text{Confluence} = \prod_{tf} \mathbb{1}_{STOCHRSI_{tf} < 0.2 \text{ or } > 0.8}$$

**交易流程**：
1. 计算N日RSI（默认14日）
2. 计算RSI的Stochastic值（默认14日周期）
3. 平滑K值（D通道，默认3日）
4. K线与D线交叉判断入场
5. 结合价格动量和成交量过滤假信号
6. 设置止损止盈

### Python代码示例
```python
"""
Stochastic RSI Strategy
随机RSI交易策略
"""
import numpy as np
import pandas as pd
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum


class SignalType(Enum):
    BUY = "buy"
    SELL = "sell"
    HOLD = "hold"


@dataclass
class TradingSignal:
    signal_type: SignalType
    strength: float  # 0-1
    price_level: float
    indicator_values: Dict[str, float]
    timestamp: Optional[datetime] = None


class RSICalculator:
    """RSI计算器"""
    
    def __init__(self, period: int = 14):
        self.period = period
        
    def calculate(self, prices: pd.Series) -> pd.Series:
        """计算RSI"""
        delta = prices.diff()
        
        gain = delta.copy()
        loss = delta.copy()
        gain[gain < 0] = 0
        loss[loss > 0] = 0
        
        # 指数移动平均
        avg_gain = gain.ewm(alpha=1/self.period, min_periods=self.period).mean()
        avg_loss = loss.ewm(alpha=1/self.period, min_periods=self.period).mean()
        
        rs = avg_gain / (avg_loss.abs() + 1e-10)
        rsi = 100 - (100 / (1 + rs))
        
        return rsi


class StochasticRSI:
    """Stochastic RSI计算器"""
    
    def __init__(self, rsi_period: int = 14, stoch_period: int = 14,
                 k_period: int = 3):
        self.rsi_period = rsi_period
        self.stoch_period = stoch_period
        self.k_period = k_period
        
    def calculate(self, prices: pd.Series) -> Tuple[pd.Series, pd.Series]:
        """计算Stochastic RSI，返回(K, D)"""
        # 计算RSI
        rsi_calc = RSICalculator(self.rsi_period)
        rsi = rsi_calc.calculate(prices)
        
        # 计算Stochastic RSI
        stoch_rsi = (rsi - rsi.rolling(self.stoch_period).min()) / \
                   (rsi.rolling(self.stoch_period).max() - 
                    rsi.rolling(self.stoch_period).min() + 1e-10)
        
        # 平滑K值（D通道）
        k = stoch_rsi.rolling(self.k_period).mean()
        
        # D值是K的SMA
        d = k.rolling(self.k_period).mean()
        
        return k, d
    
    def calculate_full(self, prices: pd.Series) -> pd.DataFrame:
        """计算完整指标"""
        k, d = self.calculate(prices)
        
        rsi_calc = RSICalculator(self.rsi_period)
        rsi = rsi_calc.calculate(prices)
        
        return pd.DataFrame({
            "rsi": rsi,
            "stoch_rsi": k,
            "k": k,
            "d": d
        })


class VolumeFilter:
    """成交量过滤器"""
    
    def __init__(self, lookback: int = 20, threshold: float = 1.5):
        self.lookback = lookback
        self.threshold = threshold
        
    def is_volume_confirmed(self, volume: pd.Series, 
                           current_volume: float) -> bool:
        """判断成交量是否确认"""
        if len(volume) < self.lookback:
            return True
            
        avg_volume = volume.iloc[-self.lookback:].mean()
        std_volume = volume.iloc[-self.lookback:].std()
        
        z_score = (current_volume - avg_volume) / (std_volume + 1e-10)
        
        return z_score > self.threshold


class MomentumFilter:
    """动量过滤器"""
    
    def __init__(self, short_period: int = 5, long_period: int = 20):
        self.short_period = short_period
        self.long_period = long_period
        
    def calculate_momentum(self, prices: pd.Series) -> pd.Series:
        """计算动量"""
        return prices.pct_change(self.short_period)
    
    def calculate_trend(self, prices: pd.Series) -> str:
        """判断趋势方向"""
        short_ma = prices.rolling(self.short_period).mean()
        long_ma = prices.rolling(self.long_period).mean()
        
        if short_ma.iloc[-1] > long_ma.iloc[-1]:
            return "UPTREND"
        elif short_ma.iloc[-1] < long_ma.iloc[-1]:
            return "DOWNTREND"
        return "RANGE"
    
    def is_momentum_confirmed(self, prices: pd.Series,
                              signal_type: SignalType) -> bool:
        """判断动量是否确认信号"""
        trend = self.calculate_trend(prices)
        momentum = self.calculate_momentum(prices).iloc[-1]
        
        if signal_type == SignalType.BUY:
            return momentum > 0 or trend == "UPTREND"
        elif signal_type == SignalType.SELL:
            return momentum < 0 or trend == "DOWNTREND"
        return True


class StochasticRSIStrategy:
    """Stochastic RSI交易策略"""
    
    def __init__(self, 
                 rsi_period: int = 14,
                 stoch_period: int = 14,
                 k_period: int = 3,
                 d_period: int = 3,
                 oversold_threshold: float = 0.2,
                 overbought_threshold: float = 0.8,
                 volume_filter: bool = True,
                 momentum_filter: bool = True):
        self.stoch_rsi = StochasticRSI(
            rsi_period, stoch_period, k_period
        )
        self.volume_filter = VolumeFilter() if volume_filter else None
        self.momentum_filter = MomentumFilter() if momentum_filter else None
        self.k_period = k_period
        self.d_period = d_period
        self.oversold_threshold = oversold_threshold
        self.overbought_threshold = overbought_threshold
        
        self.positions = {}
        self.trades = []
        self.signals = []
        
    def generate_signal(self, prices: pd.Series, 
                       volume: pd.Series) -> TradingSignal:
        """生成交易信号"""
        indicators = self.stoch_rsi.calculate_full(prices)
        k = indicators["k"].iloc[-1]
        d = indicators["d"].iloc[-1]
        rsi = indicators["rsi"].iloc[-1]
        
        # 前一期值
        k_prev = indicators["k"].iloc[-2]
        d_prev = indicators["d"].iloc[-2]
        
        # 基础信号
        signal_type = SignalType.HOLD
        strength = 0.0
        
        # 金叉买入（K从下穿越D，且在超卖区）
        if (k_prev < d_prev and k > d and k < self.oversold_threshold):
            signal_type = SignalType.BUY
            strength = 1 - k / self.oversold_threshold  # 越低越强
            
        # 死叉卖出（K从上穿越D，且在超买区）
        elif (k_prev > d_prev and k < d and k > self.overbought_threshold):
            signal_type = SignalType.SELL
            strength = (k - self.overbought_threshold) / (1 - self.overbought_threshold)
            
        # 成交量确认
        if signal_type != SignalType.HOLD and self.volume_filter:
            if not self.volume_filter.is_volume_confirmed(volume, volume.iloc[-1]):
                signal_type = SignalType.HOLD
                
        # 动量确认
        if signal_type != SignalType.HOLD and self.momentum_filter:
            if not self.momentum_filter.is_momentum_confirmed(prices, signal_type):
                signal_type = SignalType.HOLD
                
        return TradingSignal(
            signal_type=signal_type,
            strength=strength,
            price_level=prices.iloc[-1],
            indicator_values={
                "k": k,
                "d": d,
                "rsi": rsi
            }
        )
    
    def multi_timeframe_confirmation(self, 
                                    prices_daily: pd.Series,
                                    volume_daily: pd.Series,
                                    prices_weekly: pd.Series) -> TradingSignal:
        """多周期确认"""
        signal_daily = self.generate_signal(prices_daily, volume_daily)
        
        if signal_daily.signal_type == SignalType.HOLD:
            return signal_daily
            
        # 检查周线
        indicators_weekly = self.stoch_rsi.calculate_full(prices_weekly)
        k_weekly = indicators_weekly["k"].iloc[-1]
        
        # 多周期共振
        if signal_daily.signal_type == SignalType.BUY:
            if k_weekly < self.oversold_threshold:
                signal_daily.strength = min(signal_daily.strength + 0.2, 1.0)
        elif signal_daily.signal_type == SignalType.SELL:
            if k_weekly > self.overbought_threshold:
                signal_daily.strength = min(signal_daily.strength + 0.2, 1.0)
                
        return signal_daily
    
    def calculate_position_size(self, signal: TradingSignal,
                               capital: float,
                               atr: float) -> int:
        """计算仓位"""
        if signal.signal_type == SignalType.HOLD:
            return 0
            
        # ATR仓位模型
        risk_amount = capital * 0.02  # 2%风险
        stop_loss_pct = 2.0  # 2%止损
        
        if signal.indicator_values["rsi"] < 30:
            stop_loss_pct = 3.0
        elif signal.indicator_values["rsi"] > 70:
            stop_loss_pct = 3.0
            
        position_size = int(risk_amount / (atr * stop_loss_pct / 100))
        return position_size
    
    def backtest(self, prices: pd.Series, 
                volume: pd.Series,
                initial_capital: float = 100_000) -> Dict:
        """回测策略"""
        indicators = self.stoch_rsi.calculate_full(prices)
        
        capital = initial_capital
        position = 0
        entry_price = 0.0
        pnl_list = []
        trades = []
        
        for i in range(max(self.stoch_rsi.stoch_period, 
                          self.stoch_rsi.k_period * 2), len(prices)):
            
            price_window = prices.iloc[:i+1]
            volume_window = volume.iloc[:i+1]
            
            signal = self.generate_signal(price_window, volume_window)
            
            # 入场
            if signal.signal_type == SignalType.BUY and position == 0:
                position = self.calculate_position_size(
                    signal, capital, prices.iloc[i] * 0.02
                )
                entry_price = prices.iloc[i]
                
            elif signal.signal_type == SignalType.SELL and position > 0:
                # 平仓
                pnl = (prices.iloc[i] - entry_price) * position
                capital += pnl
                trades.append({
                    "entry": entry_price,
                    "exit": prices.iloc[i],
                    "pnl": pnl,
                    "signal": "BUY"
                })
                position = 0
                
            # 计算浮动盈亏
            if position > 0:
                unrealized_pnl = (prices.iloc[i] - entry_price) * position
                pnl_list.append(unrealized_pnl)
            else:
                pnl_list.append(0)
                
        return {
            "total_pnl": sum([t["pnl"] for t in trades]),
            "sharpe_ratio": np.mean([t["pnl"] for t in trades]) / 
                           np.std([t["pnl"] for t in trades]) * np.sqrt(252) if trades else 0,
            "win_rate": len([t for t in trades if t["pnl"] > 0]) / max(len(trades), 1),
            "trade_count": len(trades),
            "final_capital": capital,
            "trades": trades
        }


def generate_synthetic_price_data(n_days: int = 500) -> Tuple[pd.Series, pd.Series]:
    """生成合成价格数据"""
    np.random.seed(42)
    dates = pd.date_range(end=pd.Timestamp.now(), periods=n_days, freq="D")
    
    # 生成趋势+震荡价格
    returns = []
    for i in range(n_days):
        if i > 100 and i < 200:  # 趋势段
            ret = np.random.randn() * 0.015 + 0.002
        elif i > 300 and i < 400:  # 趋势段
            ret = np.random.randn() * 0.015 - 0.002
        else:  # 震荡
            ret = np.random.randn() * 0.01
        returns.append(ret)
        
    prices = 100 * np.exp(np.cumsum(returns))
    volume = np.random.lognormal(15, 0.5, n_days)
    
    return pd.Series(prices, index=dates), pd.Series(volume, index=dates)


if __name__ == "__main__":
    # 生成数据
    prices, volume = generate_synthetic_price_data(500)
    
    # 初始化策略
    strategy = StochasticRSIStrategy(
        rsi_period=14,
        stoch_period=14,
        k_period=3,
        volume_filter=True,
        momentum_filter=True
    )
    
    # 计算指标
    indicators = strategy.stoch_rsi.calculate_full(prices)
    
    print("最新指标值:")
    print(f"  RSI: {indicators['rsi'].iloc[-1]:.2f}")
    print(f"  Stochastic RSI: {indicators['k'].iloc[-1]:.4f}")
    print(f"  K值: {indicators['k'].iloc[-1]:.4f}")
    print(f"  D值: {indicators['d'].iloc[-1]:.4f}")
    
    # 生成信号
    signal = strategy.generate_signal(prices, volume)
    print(f"\n信号: {signal.signal_type.value}, 强度: {signal.strength:.2f}")
    
    # 回测
    result = strategy.backtest(prices, volume)
    print(f"\n回测结果:")
    print(f"  总PnL: {result['total_pnl']:.2f}")
    print(f"  交易次数: {result['trade_count']}")
    print(f"  胜率: {result['win_rate']:.2%}")
    print(f"  最终资金: {result['final_capital']:.2f}")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
